Updated color picker clear and created a color indication in title.

jorgefilipecosta opened the pull request:
## Description
This PR tries to close issue https://github.com/WordPress/gutenberg/issues/2770. By updating color picker to have a clear button in link style. And have a color indication in the panel titles.
## Testing

1. Add a paragraph block, choose a color for text and background, verify the color appears at the side of the title.
2. Repeat for the button block.
3. Collapse the color panels and verify the color indication persists.
4. Use the clear button and verify it clears the color changes. 

## Screenshots
<img width="1070" alt="screen shot 2017-10-18 at 16 19 34" src="https://user-images.githubusercontent.com/11271197/31727011-6f85643e-b420-11e7-907d-b52c6c63ff7e.png">
<img width="1076" alt="screen shot 2017-10-18 at 16 20 07" src="https://user-images.githubusercontent.com/11271197/31727018-764e821e-b420-11e7-9f81-d5466d4d7bdb.png">

jorgefilipecosta committed:
Color palette: Updated color picker clear button and created a color indication in title.

aduth commented on components/panel/body.js:
```diff
@@ -52,7 +52,14 @@ class PanelBody extends Component {
+							<span className="components-panel__body-title__text-span">
+								{ title }
+							</span>
+							{ colorValue &&
```
Is `colorValue` a concern of a general-purpose panel component? Maybe this ought to be a specialized component which extends the base panel component behaviors?

jorgefilipecosta replied:
Yes, I agree color is not a concern for a generic panel. I was not seeing the best way to extend and it looked like I needed to pass a prop. I end up making the panel body component receive an optional area to render at the side of the title. I created a PanelColor component that uses the panel body, rendering the correct area for the color.

aduth commented on components/panel/style.scss:
```diff
@@ -59,6 +59,20 @@
 	margin-bottom: 0;
+
+	.components-panel__body-title__text-span {
+    	position: relative;
```
Should be tabbed once more.

jorgefilipecosta replied:
👍 Corrected

karmatosed commented:
Visually 👍 for the design. I think you are waiting for a code review but great work getting it like the mockups.

aduth commented on components/panel/style.scss:
```diff
@@ -60,6 +60,12 @@
 }
+
+.components-panel__body-title__text-span {
+	position: relative;
```
Why is `position: relative;` needed here or below?

aduth commented on components/panel/body.js:
```diff
@@ -52,7 +52,10 @@ class PanelBody extends Component {
+							<span className="components-panel__body-title__text-span">
+								{ title }
+							</span>
+							{customTitleArea}
```
Instead of an extra `customTitleArea` prop, could the wrapping `PanelColor` component render the `title` with both the text span and color span?

```jsx
<PanelBody
	{ ...props }
	title={ [
		<span key="title">{ title }</span>,
		<span key="color" style={ { background: colorValue } }>,
	] }
/>
```

aduth commented on components/panel/body.js:
```diff
@@ -52,7 +52,10 @@ class PanelBody extends Component {
 						>
 							<Dashicon icon={ icon } />
-							{ title }
+							<span className="components-panel__body-title__text-span">
```
Would recommend to avoid including an element name in a class name, since it's incidental and can change over time, isn't representative of the semantic meaning of the element.

jorgefilipecosta commented:
Hi @aduth, I followed your sugestions and simplified the css, thank you :)

aduth reviewed (changes requested):
Small issue: When a color value is assigned, the height of the toggle is adjusted slightly. We should try to avoid this:

![resize](https://user-images.githubusercontent.com/1779930/32077053-1ac39726-ba70-11e7-85e9-24a5f375758e.gif)

mtias commented:
It might be nice to default to closed panels for text / background color now.

jorgefilipecosta commented:
I updated the PR solving the small move identified by @aduth and I made color panels hidden by default as @mtias suggested.

aduth commented on components/panel/color.js:
```diff
@@ -0,0 +1,20 @@
+import './style.scss';
+import PanelBody from './body';
+
+function PanelColor( { colorValue, title, initialOpen = false, ...props } ) {
```
@mtias Do you mean for any colored panel to default to closed, or specifically text and background color on the paragraph block?

aduth commented on components/panel/color.js:
```diff
@@ -0,0 +1,20 @@
+	return (
+		<PanelBody
+			{ ...props }
+			{ ...{ initialOpen } }
```
Aside from being a bit more difficult to read, there is a potential (minor) performance impact here on iterating an additional object to extend the props of the element, vs. the more basic `initialOpen={ initialOpen }`

See:

http://babeljs.io/repl/#?babili=false&browsers=ie%2011&build=&builtIns=true&code_lz=MYewdgzgLgBADgJxHCMC8MDeMCGAuGARgBoYAjAgJhgF8BuAKFElmHSxmAIGZbGnw0GAFMANsIC2wsFADKiYTgAm7ABQMYMADwAxECA2aOAOlOJkqGoc3ZTx7Gxq1rMKAEso4tNgDaAXWdNAHoAPgYASn5mITFJaSgAORB5BEUVDHVNXX0XWzMkFECjYG9OIs13T2FS_yLQiLogA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=es2015%2Creact&targets=&version=6.26.0

jorgefilipecosta replied:
Thank you @aduth improvement applied!

youknowriad commented on blocks/color-palette/style.scss:
```diff
@@ -3,6 +3,11 @@ $color-palette-circle-spacing: 14px;
 .blocks-color-palette {
 	margin-right: -14px;
+
+	.blocks-color-palette__clear {
```
Should we avoid nesting here? We tend to favor top-level rules when possible.

jorgefilipecosta replied:
Hi, the reason why this one is nested is for the CSS rule to have higher specificity. If the rule is not nested the margin is overwritten.

youknowriad reviewed (approved):
Nice 👍

jorgefilipecosta commented:
Hi @youknowriad thank you for the review!

jorgefilipecosta committed:
Color palette: Made paragraph block colour palette hidden by default and button block colours expanded.

jorgefilipecosta commented:
Hi, @aduth, @mtias, @youknowriad, I added the changes to make colors collapsed in the paragraph and expanded in button. In order to make things work nicely, I added a simple change in panel body that resets the state when initialOpen prop changes. The changes are in a new commit, I tested and things are looking ok.

youknowriad commented on components/panel/body.js:
```diff
@@ -36,6 +36,14 @@ class PanelBody extends Component {
 	}
 
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.initialOpen !== nextProps.initialOpen ) {
```
like its name suggest `intialOpen` should not change and if it does we should console.error. It means this panel is "uncontrolled". if we want a "controlled" panel body we can use the "opened" prop.

Can I aks why did you need this change?

jorgefilipecosta replied:
When changing blocks for example from a button to a paragraph. New props are passed, a new component is not built which makes sense so if you collapsed it in a block you can keep it collapsed when changing it. But if changing to different initial open settings without this the state keeps the same and the setting is useless.

youknowriad replied:
> a new component is not built which

I wonder why it's not the case, the block inspector of a button and a paragraph are complete separate components, I expect them to "remount" when switching between these two.

If they're not, I suppose this could create issues elsewhere. (any uncontrolled state value)

jorgefilipecosta replied:
I also expected them to remount but unfortunately, that is not the case, they just receive new props. Maybe there is an improvement that can make them remount but sure how.

aduth replied:
You can assign a unique `key` prop to trigger remounts.

youknowriad replied:
This is a really weird bug. I suspect some relation with slot-and-fill @aduth

aduth replied:
>This is a really weird bug. I suspect some relation with slot-and-fill @aduth

I don't know, it seems like React may just be recycling the same element.

aduth replied:
I expect if you give a unique key to `InspectorControls` here:

https://github.com/WordPress/gutenberg/blob/aee35468ebc4159a134672869cc6d69162e30d6a/blocks/library/paragraph/index.js#L114

Like `paragraph-inspector`... that it may resolve.

...but this sounds like a tricky implementation detail we shouldn't be expecting of block authors.

youknowriad replied:
Shouldn't we assume that `slot-fill` ensures the remounting?

youknowriad replied:
If it's a different fill, it should remount

jorgefilipecosta replied:
Hi @aduth, @youknowriad, I tried the key fix in inspector controls and I also tried adding keys to PanelColor but the components continue to be reused. It's really strange.

jorgefilipecosta replied:
It looks like react tries to reuse the components whatever it cans and although keys are a hint there is no guarantee a component will not be reused.

jorgefilipecosta replied:
It looks like the correct way is to manage state from the block components.

jorgefilipecosta commented:
Hi @aduth, @youknowriad, I researched a little bit more but the keys problem that was affecting this PR. It looks like every component we rendered in fill is "copied" to the slot by the slot-fill library, during this process the keys that the elements have are ignored and new sequential ones are assigned. That's is the reason why even setting keys on ColorPanels did not manage to make the components remount.
The children of the components passed to fill keep their keys in slot, so In the InspectorControls component, I added a div around the children. So this div will have an assigned key but children (the components we pass to InspectorControls) keep their id's. I assigned unique id's to ColorPanels in paragraph and button and now things are working as they should :)

ahmadawais commented:
Folks, I think the white color that appears from the center on focus is actually quite confusing and we should do away with it.

afercia commented on blocks/color-palette/index.js:
```diff
@@ -61,15 +61,12 @@ function ColorPalette( { colors, value, onChange } ) {
-					<span className="blocks-color-palette__clear-color-line" />
-				</button>
-			</div>
+			<button className="button-link blocks-color-palette__clear"
```
For consistency, and in the case this component gets reused within a form element, I'd consider to add a `type="button"`

aduth replied:
I also think this could be consolidated into the Button component, similar to how we have `<Button isPrimary />`, e.g. `<Button isLink />` to handle the styling and sensible defaults.

jorgefilipecosta replied:
Hi @aduth, unfortunately, color palette does not use button component I did a simple test to use it and in other cases, the design looks different. Also adding a setting isTypeButton to button component would break some tests. I will try to address this feedback in a separate PR, that makes color palette uses our button component.

afercia commented on blocks/color-palette/index.js:
```diff
@@ -61,15 +61,12 @@ function ColorPalette( { colors, value, onChange } ) {
-			</div>
+			<button className="button-link blocks-color-palette__clear"
+				onClick={ () => onChange( undefined ) }
+				aria-label={ __( 'Remove color' ) }
```
Since this button has now a visible text, the aria-label can be removed.

afercia commented on blocks/color-palette/index.js:
```diff
@@ -61,15 +61,12 @@ function ColorPalette( { colors, value, onChange } ) {
+				onClick={ () => onChange( undefined ) }
+				aria-label={ __( 'Remove color' ) }
+			>
+				{ __( 'Clear' ) }
```
Ideally, for better a11y, this text should be more specific and indicate _which_ color is going to be cleared. We should avoid to have multiple UI controls with the same name in the same page. Reading out of context, "Clear", "Clear", or for example "OK", "OK", "OK", wouldn't help users understand what a control does. Speech recognition software wouldn't be able to distinguish the controls. For example, when users will say the voice command "click clear", the software wouldn't be able to understand which button should be activated.
Ideally, they should be something like:
`Clear background color`
`Clear text color`
Not sure how to implement it, though.

afercia commented on blocks/library/button/index.js:
```diff
@@ -115,18 +115,18 @@ registerBlockType( 'core/button', {
 							onChange={ toggleClear }
 						/>
-						<PanelBody title={ __( 'Button Background Color' ) }>
+						<PanelColor title={ __( 'Button Background' ) } colorValue={ color } key="panel-color-button-1">
```
I'd recomment to don't remove the word "Color". See below. Instead, I'd consider to remove the word "Button", for consistency with the paragraphs and because it's already clear we're editing a button block.

afercia commented on blocks/library/button/index.js:
```diff
@@ -115,18 +115,18 @@ registerBlockType( 'core/button', {
-						</PanelBody>
-						<PanelBody title={ __( 'Button Text Color' ) }>
+						</PanelColor>
+						<PanelColor title={ __( 'Button Text' ) } colorValue={ textColor } key="panel-color-button-2">
```
Removing the word "Color" doesn't help users understand what this control does. "Button Text" could be the text itself, or the font, or the text size? I'd just use `Background Color` and `Text Color` for both the paragraph and the button block.

afercia commented on blocks/library/paragraph/index.js:
```diff
@@ -131,18 +131,18 @@ registerBlockType( 'core/paragraph', {
 						/>
 					</PanelBody>
-					<PanelBody title={ __( 'Background Color' ) }>
+					<PanelColor title={ __( 'Background' ) } colorValue={ backgroundColor } initialOpen={ false } key="panel-color-paragraph-1">
```
See above, I;d recommend to not remove the word "Color".

afercia commented on blocks/library/paragraph/index.js:
```diff
@@ -131,18 +131,18 @@ registerBlockType( 'core/paragraph', {
-					</PanelBody>
-					<PanelBody title={ __( 'Text Color' ) }>
+					</PanelColor>
+					<PanelColor title={ __( 'Text' ) } colorValue={ textColor } initialOpen={ false } key="panel-color-paragraph-2">
```
See above, I'd recommend to not remove the word "Color".

afercia commented on components/panel/color.js:
```diff
@@ -0,0 +1,19 @@
+			title={ [
+				<span className="components-panel__color-title" key="title">{ title }</span>,
+				colorValue && <span className="components-panel__color-area" key="color" style={ { background: colorValue } } />,
+			] }
```
I like the overall idea to show the selected color. However, this is just a visual indication. There's no text or label that can be read by software. It would be nice to find a way to add some meaningful ext, whether it's visually hidden text or an `aria-label` attribute to override the button text.

jorgefilipecosta committed:
Color palette: Changed Color Panel labels in paragraph and button blocks.

jorgefilipecosta committed:
Color palette: Added type button to clear in color palette component to be consistent with other buttons and removed an unnecessary aria-label.

jorgefilipecosta commented:
Hi @ahmadawais, thank you for your suggestion. This PR adds a color indication in the title making it easier for the user to know which color is selected, so it mitigates the problem you are referring. Although changing the way we show selected color on the pallet is something we can discuss and try to improve in the future.

Hi @afercia thank you a lot for your inputs. I addressed most of them including making color labels be "Background Color"/"Text Color" for both paragraph and button. These ones I'm not certain if for now, we should just use this labels for aria and not display color in title label because we are already showing a color cc: @mtias. Changes are in separate commits so we can easily change them or revert parts of them if needed.
Regarding your suggestions to improve aria labels for clear buttons and provide a color indication as an aria-label, I would like to address them in a separate PR  because adding them here would make this PR big and harder to review.

afercia commented:
@jorgefilipecosta thanks!
> we should just use this labels for aria and not display color in title label because we are already showing a color

Visible labels help everyone, not just assistive technologies users. I'd recommend to keep the word "Color". What was the original reasoning behind removing "Color"? Make space for the color rectangle? :) Then I'd say we're removing relevant information just for visual purposes.

ahmadawais commented:
@jorgefilipecosta 

>Although changing the way we show selected color on the pallet is something we can discuss and try to improve in the future.

That's what I want.

jorgefilipecosta commented:
Hi @aduth, @youknowriad given the merge of https://github.com/WordPress/gutenberg/pull/3404. I rebased and simplified this by removing all slot-fill related hacks. If you could have a last look at merging it would be perfect :) Then I will continue to iterate and address some more other a11y issues.

aduth commented on blocks/color-palette/index.js:
```diff
@@ -61,15 +61,11 @@ function ColorPalette( { colors, value, onChange } ) {
-					<span className="blocks-color-palette__clear-color-line" />
-				</button>
-			</div>
+			<button className="button-link blocks-color-palette__clear" type="button"
```
I'd suggest one of two conventions for applying props to an element:

Single line

```jsx
<div className="..." onClick={ () => {} }>
	Hello World
</div>
```

Multiple lines (particularly when single line exceeds 80 characters):

```jsx
<div
	className="..."
	onClick={ () => {} }
>
	Hello World
</div>
```

Currently, this is an awkward mix of single- and multi-line style.

jorgefilipecosta replied:
Corrected 👍

aduth commented on blocks/library/paragraph/index.js:
```diff
@@ -111,7 +111,7 @@ registerBlockType( 'core/paragraph', {
 			),
 			focus && (
-				<InspectorControls key="inspector">
+				<InspectorControls key="inspector" identifier="inspector-paragraph">
```
What is `identifier`?

jorgefilipecosta replied:
It was part of an experiment to solve the slot-fill keys problem. And I missed its removal it was corrected.

aduth reviewed (approved):
Merge conflicts, but otherwise 👍

jorgefilipecosta commented:
Thank you for your review @aduth, I'm resolving the merge conflicts.
